src/java.base/share/classes/jdk/internal/classfile/impl/ClassPrinterImpl.java

```diff
@@ -0,0 +1,1001 @@
+                leaf("name", evp.name().stringValue()),
+                new MapNodeImpl(FLOW, "value").with(elementValueToTree(evp.value())))));
+    }
+
+    private static Stream<ConstantDesc> convertVTIs(CodeAttribute lr, List<StackMapTableAttribute.VerificationTypeInfo> vtis) {
+        return vtis.stream().mapMulti((vti, ret) -> {
+            switch (vti) {
+                case SimpleVerificationTypeInfo s -> {
+                    switch (s) {
+                        case ITEM_DOUBLE -> {
+                            ret.accept("double");
+                            ret.accept("double2");
+                        }
+                        case ITEM_FLOAT ->
+                            ret.accept("float");
+                        case ITEM_INTEGER ->
+                            ret.accept("int");
+                        case ITEM_LONG ->  {
+                            ret.accept("long");
+                            ret.accept("long2");
+                        }
+                        case ITEM_NULL -> ret.accept("null");
+                        case ITEM_TOP -> ret.accept("?");
+                        case ITEM_UNINITIALIZED_THIS -> ret.accept("THIS");
+                    }
+                }
+                case ObjectVerificationTypeInfo o ->
+                    ret.accept(o.className().name().stringValue());
+                case UninitializedVerificationTypeInfo u ->
+                    ret.accept("UNITIALIZED @" + lr.labelToBci(u.newTarget()));
+            }
+        });
+    }
+
+    private record ExceptionHandler(int start, int end, int handler, String catchType) {}
+
+    public static MapNode modelToTree(CompoundElement<?> model, Verbosity verbosity) {
+        return switch(model) {
+            case ClassModel cm -> classToTree(cm, verbosity);
+            case FieldModel fm -> fieldToTree(fm, verbosity);
+            case MethodModel mm -> methodToTree(mm, verbosity);
+            case CodeModel com -> codeToTree((CodeAttribute)com, verbosity);
+        };
+    }
+
+    private static MapNode classToTree(ClassModel clm, Verbosity verbosity) {
+        return new MapNodeImpl(BLOCK, "class")
+                .with(leaf("class name", clm.thisClass().asInternalName()),
+                      leaf("version", clm.majorVersion() + "." + clm.minorVersion()),
+                      list("flags", "flag", clm.flags().flags().stream().map(AccessFlag::name)),
+                      leaf("superclass", clm.superclass().map(ClassEntry::asInternalName).orElse("")),
+                      list("interfaces", "interface", clm.interfaces().stream().map(ClassEntry::asInternalName)),
+                      list("attributes", "attribute", clm.attributes().stream().map(Attribute::attributeName)))
+                .with(constantPoolToTree(clm.constantPool(), verbosity))
+                .with(attributesToTree(clm.attributes(), verbosity))
```

Comment:
Is this ok? It seems we also add class attributes in the map node (see "attributes" map node).

Reply:
Not sure I get this. It seems to me that the same attributes are being printed twice. But, perhaps, the first pass only creates a list with the attribute _names_ and then there is a later pass which print the attributes in full?

Reply:
Yes, the first is just a list of attribute names, while later there might be printed more attribute details (depending on the verbosity).
Combined or dynamically changed list formats would be inconsistent in various verbosity levels, so duplicate print is preferred for better human and machine readability and consistency. 
There are other cases of duplication in favour of readability. For example some code attributes are printed standalone in a table form, as well as their individual elements are attached to the corresponding bytecode instructions.